// @testing/wdio-vbank-service/lib/utils/backends.js
"use strict";
exports.__esModule = true;
exports.backends = void 0;
var settings_1 = require("./settings");
exports.backends = {
    checkCors: checkCors,
    checkSessionLocator: checkSessionLocator,
    checkName: checkName,
    checkPort: checkPort,
    checkProtocol: checkProtocol
};
function checkCors(cors) {
    var _a = cors || {}, headers = _a.headers, methods = _a.methods;
    return cors && {
        headers: settings_1.DEFAULT_HEADERS.concat(headers || []),
        methods: settings_1.HTTP_METHODS.concat(methods || [])
    };
}
function checkSessionLocator(sessionLocator) {
    return sessionLocator || 'headers.' + settings_1.SESSION_HEADER;
}
function checkName(name, index) {
    return name || "back" + index;
}
function checkPort(backend, index) {
    var _a = backend || {}, protocol = _a.protocol, port = _a.port;
    var portChunks = (port || '').split(':');
    var isHttps = _isHttps(protocol, portChunks[1]);
    var _port = _returnPortConection(isHttps, portChunks[0]);
    return _port || ('' + (5000 + index));
}
function checkProtocol() {
    return 'http';
}
function _isHttps(protocol, isSecure) {
    if (protocol === void 0) { protocol = ''; }
    if (isSecure === void 0) { isSecure = ''; }
    return protocol.toLowerCase().trim() === 'https' || isSecure.trim() === 'true';
}
function _returnPortConection(isHttps, port) {
    return isHttps ? port + ":" + isHttps : port;
}
