// @testing/wdio-page-objects/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.containerProvider = exports.componentProvider = exports.pageProvider = void 0;
const pageProvider_1 = require("./com.bbva.arch.qe/framework/pages/pageProvider");
const containerProvider_1 = require("./com.bbva.arch.qe/framework/containers/containerProvider");
var container_1 = require("./com.bbva.arch.qe/framework/containers/container");
Object.defineProperty(exports, "ContainerObject", { enumerable: true, get: function () { return container_1.ContainerObject; } });
var component_1 = require("./com.bbva.arch.qe/framework/components/component");
Object.defineProperty(exports, "ComponentObject", { enumerable: true, get: function () { return component_1.ComponentObject; } });
var containerContext_1 = require("./com.bbva.arch.qe/framework/containers/containerContext");
Object.defineProperty(exports, "ContainerContext", { enumerable: true, get: function () { return containerContext_1.ContainerContext; } });
var componentContext_1 = require("./com.bbva.arch.qe/framework/components/componentContext");
Object.defineProperty(exports, "ComponentContext", { enumerable: true, get: function () { return componentContext_1.ComponentContext; } });
var page_1 = require("./com.bbva.arch.qe/framework/pages/page");
Object.defineProperty(exports, "DiscardablePage", { enumerable: true, get: function () { return page_1.DiscardablePage; } });
Object.defineProperty(exports, "ErrorPage", { enumerable: true, get: function () { return page_1.ErrorPage; } });
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return page_1.Page; } });
Object.defineProperty(exports, "PageObject", { enumerable: true, get: function () { return page_1.PageObject; } });
var pageContext_1 = require("./com.bbva.arch.qe/framework/pages/pageContext");
Object.defineProperty(exports, "PageContext", { enumerable: true, get: function () { return pageContext_1.PageContext; } });
__exportStar(require("./com.bbva.arch.qe/framework/utils/async-sync"), exports);
/**
 * Singleton instance of `PageProvider`. This class exposes methods to work with the PageObjects you define decorated
 * with `PageContext` and extending `PageObject`, declared in an `src\/**\/*.page.ts`
 * file.
 *
 * @see PageProvider
 */
exports.pageProvider = pageProvider_1.pageProvider;
/**
 * Singleton instance of `ComponentProvider`. This class exposes methods to work with the ContainerObjects you define decorated
 * with `ComponentContext` and extending `ComponentObject`, declared in an `src\/**\/*.component.ts`
 * file.
 *
 * @see ComponentProvider
 */
exports.componentProvider = containerProvider_1.containerProvider;
/**
 * Singleton instance of `ContainerProvider`. This class exposes methods to work with the ComponentObjects you define decorated
 * with `ContainerContext` and extending `ContainerObject`, declared in an `src\/**\/*.component.ts`
 * file.
 *
 * @see PageProvider
 */
exports.containerProvider = containerProvider_1.containerProvider;
